lib/Driver/ToolChains.cpp

```diff
@@ -1471,6 +1519,223 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+  }
+
+  bool staticExecutable = false;
+  bool staticStdlib = false;
+
+  if (context.Args.hasFlag(options::OPT_static_executable,
+                           options::OPT_no_static_executable,
+                           false)) {
+    staticExecutable = true;
+  } else if (context.Args.hasFlag(options::OPT_static_stdlib,
+                                options::OPT_no_static_stdlib,
+                                false)) {
+    staticStdlib = true;
+  }
+
+  SmallString<128> SharedRuntimeLibPath;
+  getRuntimeLibraryPath(SharedRuntimeLibPath, context.Args, *this);
+
+  SmallString<128> StaticRuntimeLibPath;
+  getRuntimeStaticLibraryPath(StaticRuntimeLibPath, context.Args, *this);
+
+  // Add the runtime library link path, which is platform-specific and found
+  // relative to the compiler.
+  if (!(staticExecutable || staticStdlib)) {
+    Arguments.push_back("-L");
+    Arguments.push_back(context.Args.MakeArgString(SharedRuntimeLibPath + "/"
+                          + getTriple().getArchName()));
+  }
+
+  // If we're cross-compiling, we likely already have some library paths set
+  // up through environment variables.
+  // Pass them through to the linker here for convenience.
+  const char *VCToolsInstallDir = ::getenv("VCToolsInstallDir");
+  const char *UniversalCRTSdkDir = ::getenv("UniversalCRTSdkDir");
+  const char *UCRTVersion = ::getenv("UCRTVersion");
+
+  StringRef tripleWinArchName;
+  switch (getTriple().getArch()) {
+    case llvm::Triple::x86:
+      tripleWinArchName = "x86";
+      break;
+    case llvm::Triple::x86_64:
+      tripleWinArchName = "x64";
+      break;
+    case llvm::Triple::arm:
+      tripleWinArchName = "arm";
+      break;
+    case llvm::Triple::aarch64:
+      tripleWinArchName = "arm64";
+      break;
+    default:
+      tripleWinArchName = getTriple().getArchName();
+      break;
+  }
+
+  StringRef VCToolsDir(VCToolsInstallDir);
+
+  if (!VCToolsDir.empty()) {
+    Arguments.push_back("-L");
+    Arguments.push_back(context.Args.MakeArgString(llvm::Twine(VCToolsInstallDir) 
+        + "/Lib/" + tripleWinArchName));
+  }
+
+  StringRef UCRTDir(UniversalCRTSdkDir);
+  StringRef UCRTVer(UCRTVersion);
+
+  if (!UCRTDir.empty() && !UCRTVer.empty()) {
+      Arguments.push_back("-L");
+      Arguments.push_back(context.Args.MakeArgString(llvm::Twine(UCRTDir) 
+          + "/Lib/" + llvm::Twine(UCRTVer) + "/ucrt/" + tripleWinArchName));
+      Arguments.push_back("-L");
+      Arguments.push_back(context.Args.MakeArgString(llvm::Twine(UCRTDir) 
+          + "/Lib/" + llvm::Twine(UCRTVer) + "/um/" + tripleWinArchName));
+  }
+
+  addPrimaryInputsOfType(Arguments, context.Inputs, types::TY_Object);
+  addInputsOfType(Arguments, context.InputActions, types::TY_Object);
+
+  for (const Arg *arg : context.Args.filtered(options::OPT_F,
+                                              options::OPT_Fsystem)) {
+    if (arg->getOption().matches(options::OPT_Fsystem))
+      Arguments.push_back("-iframework");
```

Comment:
Can we ever get frameworks being passed to the linker in a Windows context?

Reply:
The only thing that is interesting about a framework is the fact that they have a particular layout.  I don't see a strong reason to not forward those along to clang and have clang just transliterate them into the appropriate `-I` and `-L` options.